services/wallet/market/market.go

```diff
@@ -133,9 +143,86 @@ func (pm *Manager) FetchTokenMarketValues(symbols []string, currency string) (ma
 	}
 
 	marketValues := result.(map[string]thirdparty.TokenMarketValues)
+	pm.updateMarketCache(currency, marketValues)
 	return marketValues, nil
 }
 
+func (pm *Manager) GetCachedTokenMarketValues() MarketValuesPerCurrencyAndToken {
+	pm.marketCacheLock.RLock()
+	defer pm.marketCacheLock.RUnlock()
+	return pm.marketCache
+}
+
+func (pm *Manager) GetOrFetchTokenMarketValues(symbols []string, currency string, maxAgeInSeconds int64) (map[string]thirdparty.TokenMarketValues, error) {
+	symbolsToFetchMap := make(map[string]bool)
+	symbolsToFetch := make([]string, 0, len(symbols))
+
+	now := time.Now().Unix()
+	tokenMarketValuesCache, ok := pm.GetCachedTokenMarketValues()[currency]
+	if !ok {
+		return pm.FetchTokenMarketValues(symbols, currency)
+	}
+
+	for _, symbol := range symbols {
+		marketValueSnapshot, found := tokenMarketValuesCache[symbol]
+		if !found {
+			if !symbolsToFetchMap[symbol] {
+				symbolsToFetchMap[symbol] = true
+				symbolsToFetch = append(symbolsToFetch, symbol)
+			}
+			continue
+		}
+		if now-marketValueSnapshot.UpdatedAt > maxAgeInSeconds {
+			if !symbolsToFetchMap[symbol] {
+				symbolsToFetchMap[symbol] = true
+				symbolsToFetch = append(symbolsToFetch, symbol)
+			}
+			continue
+		}
+	}
```

Comment:
Here we're determining which tokens need their market data requested. If they don't have any data in the cache, then we must request the data. And if they have stale data in the cache, then we also will request the data. Otherwise, we'll use whatever is in the cache.